circle.go

```diff
@@ -72,11 +73,75 @@ type CircleArtifact struct {
 }
 
 type CircleBuild struct {
+	BuildNum                uint32         `json:"build_num"`
 	Parallel                uint8          `json:"parallel"`
 	PreviousSuccessfulBuild PreviousBuild  `json:"previous_successful_build"`
 	QueuedAt                types.NullTime `json:"queued_at"`
+	RepoName                string         `json:"reponame"` // "go"
 	Steps                   []Step         `json:"steps"`
+	VCSType                 string         `json:"vcs_type"` // "github", "bitbucket"
 	UsageQueuedAt           types.NullTime `json:"usage_queued_at"`
+	Username                string         `json:"username"` // "golang"
+}
+
+// Failures returns an array of (buildStep, containerID) integers identifying
+// the IDs of container/build step pairs that failed.
+func (cb *CircleBuild) Failures() [][2]int {
+	failures := make([][2]int, 0)
+	for i, step := range cb.Steps {
+		for j, action := range step.Actions {
+			if action.Failed() {
+				failures = append(failures, [...]int{i, j})
+			}
+		}
+	}
+	return failures
+}
+
+type CircleOutput struct {
+	Message string    `json:"message"`
+	Time    time.Time `json:"time"`
+	Type    string    `json:"type"`
+}
+
+type CircleOutputs []*CircleOutput
+
+func (cb *CircleBuild) FailureTexts(ctx context.Context) ([]string, error) {
+	group, errctx := errgroup.WithContext(ctx)
+	// todo this is not great design
+	token, err := getToken(cb.Username)
+	if err != nil {
+		return nil, err
+	}
+	failures := cb.Failures()
+	results := make([]string, len(failures))
+	for i, failure := range failures {
+		failure := failure
+		i := i
+		// https://circleci.com/api/v1.1/project/github/segmentio/terracode-bootstrap/11/output/9/0
```

Comment:
Oh "Project not found" only because it's a private repository :( I think it should work if you replace terracode-bootstrap with a public CircleCI repo?